## Clean & Analyze Employee Exits Surveys

In this guided project, we are working with exit surveys from employees of the Department of Education, Training, and Employment (DETE) and the Technical and Further Education (TAFE) institute in Queensland, Australia.

In this project we focus on providing insight about the following questions:
+ Are employees who only worked for the institutes for a short period of time resigning due to some kind of dissatisfaction? What about employees who have been there longer?
+ Are younger employees resigning due to some kind of dissatisfaction? What about older employees?

Some of the data columns we will be working with are defined as:
From the `dete_survey.csv`: 

+ `ID`: An id used to identify the participant of the survey
+ `SeparationType`: The reason why the person's employment ended
+ `Cease Date`: The year or month the person's employment ended
+ `DETE Start Date`: The year the person began employment with the DETE

From the `tafe_survey.csv`:

+ `Record ID`: An id used to identify the participant of the survey
+ `Reason for ceasing employment`: The reason why the person's employment ended
+ `LengthofServiceOverall. Overall Length of Service at Institute (in years)`: The length of the person's employment (in years)

In [1]:
# Import libraries
import numpy as np
import pandas as pd

In [2]:
# Read the csv files
dete_survey = pd.read_csv('dete_survey.csv')
tafe_survey = pd.read_csv('tafe_survey.csv')

In [3]:
# Print some info about the datasets
dete_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 822 entries, 0 to 821
Data columns (total 56 columns):
ID                                     822 non-null int64
SeparationType                         822 non-null object
Cease Date                             822 non-null object
DETE Start Date                        822 non-null object
Role Start Date                        822 non-null object
Position                               817 non-null object
Classification                         455 non-null object
Region                                 822 non-null object
Business Unit                          126 non-null object
Employment Status                      817 non-null object
Career move to public sector           822 non-null bool
Career move to private sector          822 non-null bool
Interpersonal conflicts                822 non-null bool
Job dissatisfaction                    822 non-null bool
Dissatisfaction with the department    822 non-null bool
Physical work environ

In [4]:
dete_survey.head()

ID                    SeparationType Cease Date DETE Start Date  \
0   1             Ill Health Retirement    08/2012            1984   
1   2  Voluntary Early Retirement (VER)    08/2012      Not Stated   
2   3  Voluntary Early Retirement (VER)    05/2012            2011   
3   4         Resignation-Other reasons    05/2012            2005   
4   5                    Age Retirement    05/2012            1970   

  Role Start Date                                      Position  \
0            2004                                Public Servant   
1      Not Stated                                Public Servant   
2            2011                               Schools Officer   
3            2006                                       Teacher   
4            1989  Head of Curriculum/Head of Special Education   

  Classification              Region                      Business Unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     Employment Status  ...   Kept informed  Wellness programs  \
0  Permanent Full-time  ...               N                  N   
1  Permanent Full-time  ...               N                  N   
2  Permanent Full-time  ...               N                  N   
3  Permanent Full-time  ...               A                  N   
4  Permanent Full-time  ...               N                  A   

   Health & Safety  Gender          Age  Aboriginal  Torres Strait  South Sea  \
0                N    Male        56-60         NaN            NaN        NaN   
1                N    Male        56-60         NaN            NaN        NaN   
2                N    Male  61 or older         NaN            NaN        NaN   
3                A  Female        36-40         NaN            NaN        NaN   
4                M  Female  61 or older         NaN            NaN        NaN   

   Disability  NESB  
0         NaN   Yes  
1         NaN   NaN  
2         NaN   NaN  
3         NaN   NaN  
4         NaN   NaN  

[5 rows x 56 columns]

In [5]:
# info about the tafe_survey
tafe_survey.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 72 columns):
Record ID                                                                                                                                                        702 non-null float64
Institute                                                                                                                                                        702 non-null object
WorkArea                                                                                                                                                         702 non-null object
CESSATION YEAR                                                                                                                                                   695 non-null float64
Reason for ceasing employment                                                                                                                                    701 non-

In [6]:
tafe_survey.head()

Record ID                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   WorkArea  CESSATION YEAR Reason for ceasing employment  \
0  Non-Delivery (corporate)          2010.0              Contract Expired   
1  Non-Delivery (corporate)          2010.0                    Retirement   
2       Delivery (teaching)          2010.0                    Retirement   
3  Non-Delivery (corporate)          2010.0                   Resignation   
4       Delivery (teaching)          2010.0                   Resignation   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

                                     ...                                     \
0                                    ...                                      
1                                    ...                                      
2                                    ...                                      
3                                    ...                                      
4                                    ...                                      

  Workplace. Topic:Does your workplace promote a work culture free from all forms of unlawful discrimination?  \
0                                                Yes                                                            
1                                                Yes                                                            
2                                                Yes                                                            
3                                                Yes                                                            
4                                                Yes                                                            

  Workplace. Topic:Does your workplace promote and practice the principles of employment equity?  \
0                                                Yes                                               
1                                                Yes                                               
2                                                Yes                                               
3                                                Yes                                               
4                                                Yes  

In [7]:
# check if there are null values in the columns of the dataset
dete_survey.isnull().sum()

ID                                       0
SeparationType                           0
Cease Date                               0
DETE Start Date                          0
Role Start Date                          0
Position                                 5
Classification                         367
Region                                   0
Business Unit                          696
Employment Status                        5
Career move to public sector             0
Career move to private sector            0
Interpersonal conflicts                  0
Job dissatisfaction                      0
Dissatisfaction with the department      0
Physical work environment                0
Lack of recognition                      0
Lack of job security                     0
Work location                            0
Employment conditions                    0
Maternity/family                         0
Relocation                               0
Study/Travel                             0
Ill Health 

In [8]:
# check if there are null values in the tafe_survey
tafe_survey.isnull().sum()

Record ID                                                                                                                                                          0
Institute                                                                                                                                                          0
WorkArea                                                                                                                                                           0
CESSATION YEAR                                                                                                                                                     7
Reason for ceasing employment                                                                                                                                      1
Contributing Factors. Career Move - Public Sector                                                                                                                265
Contributi

In [9]:
dete_survey.shape

(822, 56)

In [10]:
tafe_survey.shape

(702, 72)

From observing the data, the formats of both surveys are very different. 
`dete_survey`:

+ Age is listed as a range
+ There appears to be some columns with >90% NaN values 
+ Shape of dataset is 822 rows and 56 columns

`tafe_survey`:

+ in the age column, some have null values or 2 numbers represented
+ the responses are 'Yes' 'No' to survey
+ there are - for no response with a mix of NaN
+ Shape of dataset is 702 rows and 72 columns

### Summary of observations from the datasets

+ The `dete_survey` dataframe contains `Not Stated` values that indicate values are missing, but they aren't represented as `NaN`.
+ Both the surveys dataframes contain many columns that we don't need to complete our analysis.
+ Each dataframe contains many of the same columns, but the column names are different.
+ There are multiple columns/answers that indicate an employee resigned because they were dissatisfied. 

To start we will use the `pd.read_csv()` to specify which values should be represented as `NaN`. We use this function to fix the missing values first. Then, we'll drop columns we know we don't need for our analysis

In [11]:
# read the dete_survey agan and Not Stated values as NaN
dete_survey = pd.read_csv('dete_survey.csv', na_values='Not Stated')

In [12]:
# Drop columns that are not needed
dete_survey_updated = dete_survey.drop(dete_survey.columns[28:49], axis=1)

In [13]:
# To check number of columns
dete_survey_updated.shape

(822, 35)

In [14]:
# Drop columns from the tafe_survey
tafe_survey_updated = tafe_survey.drop(tafe_survey.columns[17:66], axis=1)

In [15]:
# To check columns dropped
tafe_survey_updated.shape

(702, 23)

## Cleaning Column Names

Each dataframe contains many of the same columns, but the column names are different. The columns we'd like to use for our analysis:

| dete_survey  | tafe_survey  | Definition  |
|---|---|---|
|  ID |  Record ID |  An id used to identify the participant of the survey |
|  SeparationType | Reason for ceasing employment  |  The reason why the participant's employment ended |
|  Cease Date |CESSATION YEAR   |  The year or month the participant's employment ended |
|  DETE Start Date |   | The year the participant began employment with the DETE  |
|   |  LengthofServiceOverall. Overall Length of Service (in years) | The length of the person's employment (in years)  |
|  Age |  CurrentAge.Current Age | The age of the participant  |
|  Gender | Gender. What is your Gender  | The gender of the participant  |

We will need to standardize the column names. We've done this before in an exercise
`happiness2017.columns = happiness2017.columns.str.replace('.', ' ').str.replace('\s+', ' ').str.strip().str.upper()`

In [16]:
# Rename the columns in dete survey
dete_survey_updated 

ID                        SeparationType Cease Date  DETE Start Date  \
0      1                 Ill Health Retirement    08/2012           1984.0   
1      2      Voluntary Early Retirement (VER)    08/2012              NaN   
2      3      Voluntary Early Retirement (VER)    05/2012           2011.0   
3      4             Resignation-Other reasons    05/2012           2005.0   
4      5                        Age Retirement    05/2012           1970.0   
5      6             Resignation-Other reasons    05/2012           1994.0   
6      7                        Age Retirement    05/2012           1972.0   
7      8                        Age Retirement    05/2012           1988.0   
8      9             Resignation-Other reasons    07/2012           2009.0   
9     10            Resignation-Other employer       2012           1997.0   
10    11                        Age Retirement       2012           1999.0   
11    12  Resignation-Move overseas/interstate       2012           2009.0   
12    13             Resignation-Other reasons       2012           1998.0   
13    14                        Age Retirement       2012           1967.0   
14    15            Resignation-Other employer       2012           2007.0   
15    16      Voluntary Early Retirement (VER)       2012           1995.0   
16    17             Resignation-Other reasons       2012              NaN   
17    18                        Age Retirement       2012           1996.0   
18    19                        Age Retirement       2012           2006.0   
19    20                        Age Retirement       2012           1989.0   
20    21            Resignation-Other employer       2012           1982.0   
21    22             Resignation-Other reasons       2012           1980.0   
22    23             Resignation-Other reasons       2012           1997.0   
23    24             Resignation-Other reasons       2012           1973.0   
24    25                        Age Retirement       2012           1981.0   
25    26             Resignation-Other reasons       2012           1995.0   
26    27                        Age Retirement       2012           1974.0   
27    28            Resignation-Other employer       2012           2005.0   
28    29                        Age Retirement       2012           1989.0   
29    30                        Age Retirement       2012           1975.0   
..   ...                                   ...        ...              ...   
792  794                                 Other        NaN           1996.0   
793  795                 Ill Health Retirement    12/2013           2003.0   
794  796  Resignation-Move overseas/interstate    11/2013              NaN   
795  797                        Age Retirement        NaN           1977.0   
796  798                        Age Retirement        NaN           1970.0   
797  799  Resignation-Move overseas/interstate        NaN           2000.0   
798  800  Resignation-Move overseas/interstate        NaN           1995.0   
799  801                                 Other        NaN           1981.0   
800  802                                 Other    01/2014           2011.0   
801  803                 Ill Health Retirement    01/2014           1972.0   
802  804  Resignation-Move overseas/interstate    12/2013              NaN   
803  805            Resignation-Other employer    01/2014           2004.0   
804  806            Resignation-Other employer    01/2014           2008.0   
805  807                        Age Retirement        NaN           1990.0   
806  808            Resignation-Other employer    12/2013           2005.0   
807  809             Resignation-Other reasons    12/2013           2004.0   
808  810             Resignation-Other reasons    12/2013           2010.0   
809  811                        Age Retirement    01/2014           1979.0   
810  812                        Age Retirement        NaN           1972.0   
811  813                    

In [17]:
# When changing the columns make sure you use dataframe.columns
dete_survey_updated.columns = dete_survey_updated.columns.str.replace('\s+','_').str.strip().str.lower()

In [19]:
# Print the column names
dete_survey_updated.columns

Index(['id', 'separationtype', 'cease_date', 'dete_start_date',
       'role_start_date', 'position', 'classification', 'region',
       'business_unit', 'employment_status', 'career_move_to_public_sector',
       'career_move_to_private_sector', 'interpersonal_conflicts',
       'job_dissatisfaction', 'dissatisfaction_with_the_department',
       'physical_work_environment', 'lack_of_recognition',
       'lack_of_job_security', 'work_location', 'employment_conditions',
       'maternity/family', 'relocation', 'study/travel', 'ill_health',
       'traumatic_incident', 'work_life_balance', 'workload',
       'none_of_the_above', 'gender', 'age', 'aboriginal', 'torres_strait',
       'south_sea', 'disability', 'nesb'],
      dtype='object')

In [20]:
# Rename the columns in tafe survey
tafe_survey_updated = tafe_survey_updated.rename(columns={'Record ID':'id', 'CESSATION YEAR':'cease_date',
                                                         'Reason for ceasing employment': 'separationtype',
                                                         'Gender. What is your Gender?': 'gender',
                                                         'CurrentAge. Current Age': 'age',
                                                         'Employment Type. Employment Type': 'employment_status',
                                                         'Classification. Classifciation': 'position',
                                                         'LengthofServiceOverall. Overall Length of Service at Institute (in years)': 'institute_service',
                                                         'LengthofServiceCurrent. Length of Service at current workplace (in years)': 'role_service'}
                                                )

In [21]:
dete_survey_updated.head()

id                    separationtype cease_date  dete_start_date  \
0   1             Ill Health Retirement    08/2012           1984.0   
1   2  Voluntary Early Retirement (VER)    08/2012              NaN   
2   3  Voluntary Early Retirement (VER)    05/2012           2011.0   
3   4         Resignation-Other reasons    05/2012           2005.0   
4   5                    Age Retirement    05/2012           1970.0   

   role_start_date                                      position  \
0           2004.0                                Public Servant   
1              NaN                                Public Servant   
2           2011.0                               Schools Officer   
3           2006.0                                       Teacher   
4           1989.0  Head of Curriculum/Head of Special Education   

  classification              region                      business_unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     employment_status  ...   work_life_balance  workload  none_of_the_above  \
0  Permanent Full-time  ...               False     False               True   
1  Permanent Full-time  ...               False     False              False   
2  Permanent Full-time  ...               False     False               True   
3  Permanent Full-time  ...               False     False              False   
4  Permanent Full-time  ...                True     False              False   

   gender          age  aboriginal  torres_strait  south_sea  disability  nesb  
0    Male        56-60         NaN            NaN        NaN         NaN   Yes  
1    Male        56-60         NaN            NaN        NaN         NaN   NaN  
2    Male  61 or older         NaN            NaN        NaN         NaN   NaN  
3  Female        36-40         NaN            NaN        NaN         NaN   NaN  
4  Female  61 or older         NaN            NaN        NaN         NaN   NaN  

[5 rows x 35 columns]

In [22]:
tafe_survey_updated.head()

id                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   WorkArea  cease_date    separationtype  \
0  Non-Delivery (corporate)      2010.0  Contract Expired   
1  Non-Delivery (corporate)      2010.0        Retirement   
2       Delivery (teaching)      2010.0        Retirement   
3  Non-Delivery (corporate)      2010.0       Resignation   
4       Delivery (teaching)      2010.0       Resignation   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

      ...      Contributing Factors. Study Contributing Factors. Travel  \
0     ...                              NaN                          NaN   
1     ...                                -                       Travel   
2     ...                                -                            -   
3     ...                                -                       Travel   
4     ...                                -                            -   

  Contributing Factors. Other Contributing Factors. NONE  gender     age  \
0                         NaN                        NaN  Female  26  30   
1                           -                          -     NaN     NaN   
2                           -                       NONE     NaN     NaN   
3                           -                          -     NaN     NaN   
4                           -                          -    Male  41  45   

     employment_status Classification. Classification institute_service  \
0  Temporary Full-time            Administration (AO)               1-2   
1                  NaN                            NaN               NaN   
2                  NaN                            NaN               NaN   
3                  NaN                            NaN               NaN   
4  Permanent Full-time        Teacher (including LVT)               3-4   

  role_service  
0          1-2  
1          NaN  
2          NaN  
3          NaN  
4          3-4  

[5 rows x 23 columns]

## Filter the Data

